README.md
# `libftdi1-sys`
`libftdi1-sys` is a crate providing Rust bindings to the C library
[`libftdi1`](https://www.intra2net.com/en/developer/libftdi/index.php).

# Prerequisites
This crate requires `libftdi1` version 1.4 (August 2017) or later to be available as a system library
that can be found with `pkg-config` (everywhere except windows/MSVC) or `vcpkg` (windows/MSVC)
unless you activate the `vendored` feature.

By default the crate uses pregenerated bindings which should be fine in most cases.
In special cases the bindings can be regenerated using the `bindgen` feature.

Regenerating bindings has an additional requirement that 
  `libclang` must be installed and visible on your path. If you are using a
  `gcc`-toolchain and don't want to install the entirity of LLVM just for
  `libclang`, you can use the following procedure (using a Debian-based
  ARM system as an example):

  ```
  sudo apt-get install libclang-dev
  export LIBCLANG_PATH=/usr/lib/llvm-7/lib
  export C_INCLUDE_PATH=/usr/lib/gcc/arm-linux-gnueabihf/8/include
  cargo build
  ```

# MSRV
The Minimum Supported Rust Version (MSRV) is stable `1.34`.
The `bindgen` feature is developer-oriented and does not provide any MSRV guarantees.

# Features
* `libusb1-sys`: depend on `libusb1-sys` and use real `libusb` types instead of placeholders.
This makes it possible to interact directly with the underlying `libusb` structures.
* `vendored`: build a custom copy of `libftdi` instead of using the system one.
Note that this includes LGPL code in your build.
* `bindgen`: Generate bindings to `libftdi` at compile time.

# Contributing

This crate is expected to be passively maintained, not actively updated.
That being said, if you want to update binding generation,
you can set the `LIBFTDI1_SYS_DEVEL` environment variable with any value
so that an updated binding file is written in `src` alongside the original one
enabling you to compare them and copy it over the old one to update.


CHANGELOG.md
# 1.1.3 (2023-12-14)

Changed:

- use the rustonomicon-recommended extern type pattern without `libusb1-sys` (#19)

# 1.1.2 (2022-02-13)

Changed:

- allow `libusb1-sys` versions up to `0.6.*`

# 1.1.1 (2021-08-12)

Fixed:

- fix `bindgen` feature compatibility with `libftdi` before 1.5

# 1.1.0 (2021-08-12)

BREAKING:

Changed the SIO constants used for flow control modes to be signed.
Their only real use is with the matching function, which takes `c_int`,
while they used to be `u32`. Formally speaking this is a breaking change, but:
- I've manually verified all `crates.io` revdeps to be okay with that (they don't use them);
- this change removes a mandatory and completely useless type conversion.

Changed:

- allow `libusb` versions up to `0.5.*`
- bump dependencies on `cfg-if` and `bindgen`
  - as a side effect, enums now use the `c_uint` type, not `u32` directly
- deprecate `SIO_RESET_PURGE_{RX,TX}` (deprecated in `libftdi` 1.5)


build.rs
// SPDX-License-Identifier: MIT

use std::env;
use std::path::PathBuf;

cfg_if::cfg_if! {
    if #[cfg(feature = "bindgen")] {
        use bindgen;
    }
}

cfg_if::cfg_if! {
    if #[cfg(all(windows, target_env="msvc"))] {
        fn find_library() -> Result<vcpkg::Library, vcpkg::Error> {
            vcpkg::find_package("libftdi1")
        }
    } else {
        fn find_library() -> Result<pkg_config::Library, pkg_config::Error> {
            pkg_config::Config::new().atleast_version("1.4").probe("libftdi1")
        }
    }
}

fn main() {
    if cfg!(feature = "vendored") {
        build_source();
    } else {
        if cfg!(feature = "libusb1-sys") {
            match env::var("DEP_USB_1.0_STATIC") {
                Ok(ref val) if val == "1" => {
                    panic!("libftdi1-sys can use static libusb1-sys with `vendored` feature only");
                }
                _ => {}
            }
        }

        if let Err(err) = find_library() {
            println!(
                "cargo:warning={}: {}",
                "Config for libftdi1 not found, falling back to default search path", err,
            );
            println!("cargo:rustc-link-lib=dylib=ftdi1");
        }
    }

    cfg_if::cfg_if! {
        if #[cfg(feature = "bindgen")] {

            #[derive(Debug)]
            struct Callbacks;

            use bindgen::callbacks::{IntKind, MacroParsingBehavior, ParseCallbacks};

            impl ParseCallbacks for Callbacks {
                fn will_parse_macro(&self, name: &str) -> MacroParsingBehavior {
                    static OVERRIDE_IGNORE: &[&str] = &[
                        "SIO_RESET_PURGE_RX",
                        "SIO_RESET_PURGE_TX",
                    ];

                    if OVERRIDE_IGNORE.contains(&name) {
                        MacroParsingBehavior::Ignore
                    } else {
                        MacroParsingBehavior::Default
                    }
                }
                fn int_macro(&self, name: &str, _value: i64)-> Option<IntKind> {
                    static OVERRIDE_SIGNED: &[&str] = &[
                        "SIO_DISABLE_FLOW_CTRL",
                        "SIO_RTS_CTS_HS",
                        "SIO_DTR_DSR_HS",
                        "SIO_XON_XOFF_HS",
                    ];

                    if OVERRIDE_SIGNED.contains(&name) {
                        Some(IntKind::Int)
                    } else {
                        None
                    }
                }
            }

            fn bindings_builder() -> bindgen::Builder {
                bindgen::Builder::default()
                    .header(header_path())
                    .default_enum_style(bindgen::EnumVariation::NewType {
                        is_bitfield: false,
                        is_global: false,
                    })
                    .parse_callbacks(Box::new(Callbacks))
                    .allowlist_function("ftdi_.*")
                    .allowlist_type("ftdi_.*")
                    .generate_comments(false)
                    .allowlist_var("SIO_.*")
                    .no_copy("ftdi_eeprom")
                    .no_copy("ftdi_device_list")
                    .no_copy("ftdi_context")
                    .no_copy("ftdi_transfer_control")
                    .blocklist_type("timeval")
                    .blocklist_type("libusb_.*")
                    .blocklist_type("__.*")
                    .raw_line("pub type timeval = libc::timeval;")
                    // The following use too much macros to parse correctly in libftdi 1.5
                    // Inject manually, they're ignored in callbacks
                    .raw_line("#[deprecated] pub const SIO_RESET_PURGE_RX: u32 = 1;")
                    .raw_line("#[deprecated] pub const SIO_RESET_PURGE_TX: u32 = 2;")
            }

            let out_path = PathBuf::from(env::var("OUT_DIR").unwrap());
            bindings_builder()
                .generate()
                .expect("Unable to generate bindings")
                .write_to_file(out_path.join("bindings.rs"))
                .expect("Couldn't write bindings!");

            println!("cargo:rerun-if-env-changed=LIBFTDI1_SYS_DEVEL");
            if env::var("LIBFTDI1_SYS_DEVEL").is_ok() {
                bindings_builder()
                    .layout_tests(false)
                    .generate()
                    .expect("Unable to generated bindings without tests")
                    .write_to_file("src/pregenerated.rs.updated")
                    .expect("Couldn't update pregenerated bindings!");
            }
        } else {
            // Anything not depending on bindgen feature goes here.
        }
    }
}

cfg_if::cfg_if! {
    if #[cfg(feature = "vendored")] {
        fn build_source() {
            let source = PathBuf::from(env::var("DEP_FTDI1_SOURCE_SOURCE_DIR")
                .expect("no source found"));

            let include_paths = link_and_get_include_paths();

            let mut libftdi = cc::Build::new();
            for path in include_paths {
                libftdi.include(path);
            }
            libftdi
                .files(&[source.join("ftdi.c"), source.join("ftdi_stream.c")])
                .include(source)
                .compile("ftdi");

        }

        fn header_path() -> String {
            let source = PathBuf::from(env::var("DEP_FTDI1_SOURCE_SOURCE_DIR")
                .expect("no source found"));
            source.join("ftdi.h").to_str().unwrap().to_owned()
        }
    } else {
        fn build_source() {}

        fn header_path() -> String {
            "wrapper.h".into()
        }
    }
}

cfg_if::cfg_if! {
    if #[cfg(feature = "libusb1-sys")] {
        fn link_and_get_include_paths() -> Vec<PathBuf> {
            vec![
                PathBuf::from(env::var("DEP_USB_1.0_INCLUDE")
                    .expect("libusb is required for libftdi"))
            ]
        }
    } else if #[cfg(all(windows, target_env="msvc"))] {
        fn link_and_get_include_paths() -> Vec<PathBuf> {
            let libusb = vcpkg::find_package("libusb-1.0")
                .expect("libusb is required for libftdi");
            libusb.include_paths
        }
    } else {
        fn link_and_get_include_paths() -> Vec<PathBuf> {
            let libusb = pkg_config::probe_library("libusb-1.0")
                .expect("libusb is required for libftdi");
            libusb.include_paths
        }
    }
}
